core/station/impl/src/services/external_canister.rs

```diff
@@ -72,4 +128,357 @@ impl ExternalCanisterService {
+                EnsureExternalCanister::is_external_canister(opts.canister_id)?;
+
+                self.check_unique_canister_id(&opts.canister_id, None)?;
+
+                opts.canister_id
+            }
+        };
+
+        let external_canister = ExternalCanisterMapper::from_create_input(canister_id, input);
+
+        external_canister.validate()?;
+
+        // todo: add permissions and request policies handling
+
+        self.external_canister_repository
+            .insert(external_canister.to_key(), external_canister.clone());
+
+        Ok(external_canister)
+    }
+
+    /// Edits an external canister's settings.
+    pub fn edit_external_canister(
+        &self,
+        id: &ExternalCanisterId,
+        input: ConfigureExternalCanisterSettingsInput,
+    ) -> ServiceResult<ExternalCanister> {
+        let mut external_canister = self.get_external_canister(id)?;
+
+        external_canister.update_with(input);
+        external_canister.validate()?;
+
+        // todo: add permissions and request policies handling
+
+        self.external_canister_repository
+            .insert(external_canister.to_key(), external_canister.clone());
+
+        Ok(external_canister)
+    }
+
+    /// Adds cycles to the external canister, the cycles are taken from the station's balance.
+    pub async fn top_up_canister(&self, canister_id: Principal, cycles: u128) -> ServiceResult<()> {
+        if let Err((err_code, err_msg)) =
+            deposit_cycles(CanisterIdRecord { canister_id }, cycles).await
+        {
+            Err(ExternalCanisterError::Failed {
+                reason: format!(
+                    "Failed to top up canister {} with {} cycles, code: {:?} and reason: {:?}",
+                    canister_id.to_text(),
+                    cycles,
+                    err_code,
+                    err_msg
+                ),
+            })?;
+        }
+
+        Ok(())
+    }
+
+    /// Only deletes the external canister from the system.
+    pub fn soft_delete_external_canister(
+        &self,
+        id: &ExternalCanisterId,
+    ) -> ServiceResult<ExternalCanister> {
+        let external_canister = self.get_external_canister(id)?;
+        self.external_canister_repository
+            .remove(&external_canister.to_key());
+
+        // todo: remove permissions and request policies
+
+        Ok(external_canister)
+    }
+
+    /// Deletes an external canister from the system, as well as from the subnet.
+    pub async fn hard_delete_external_canister(
+        &self,
+        id: &ExternalCanisterId,
+    ) -> ServiceResult<ExternalCanister> {
+        let external_canister = self.get_external_canister(id)?;
+
+        // Deleting a canister requires the canister to be stopped first.
+        //
+        // See https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-delete_canister
+        if let Err((err_code, err_msg)) = stop_canister(CanisterIdRecord {
```

Comment:
does this return an error if the canister is already stopped? maybe still go through with trying to delete the canister even if it errors out?

Reply:
i think it would error out, i'll update to continue and just print the error if stop fails.